backends/ebpf/psa/externs/ebpfPsaRegister.cpp

```diff
@@ -0,0 +1,230 @@
+/*
+Copyright 2022-present Orange
+Copyright 2022-present Open Networking Foundation
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "ebpfPsaRegister.h"
+#include "backends/ebpf/psa/ebpfPsaControl.h"
+
+namespace EBPF {
+
+EBPFRegisterPSA::EBPFRegisterPSA(const EBPFProgram *program,
+                                 cstring instanceName, const IR::Declaration_Instance *di,
+                                 CodeGenInspector *codeGen) : EBPFTableBase(program,
+                                                                            instanceName,
+                                                                            codeGen) {
+    CHECK_NULL(di);
+    if (!di->type->is<IR::Type_Specialized>()) {
+        ::error(ErrorType::ERR_MODEL, "Missing specialization: %1%", di);
+        return;
+    }
+    auto ts = di->type->to<IR::Type_Specialized>();
+
+    if (ts->arguments->size() != 2) {
+        ::error(ErrorType::ERR_MODEL, "Expected a type specialized with two arguments: %1%", ts);
+        return;
+    }
+
+    this->keyArg = ts->arguments->at(1);
+    this->valueArg = ts->arguments->at(0);
+    this->keyType = EBPFTypeFactory::instance->create(keyArg);
+    this->valueType = EBPFTypeFactory::instance->create(valueArg);
+
+    if (auto wt = dynamic_cast<IHasWidth *>(this->keyType)) {
+        unsigned keyWidth = wt->widthInBits();
+        // For keys <= 32 bit register is based on array map,
+        // otherwise we use hash map
+        useArrayMap = (keyWidth > 0 && keyWidth <= 32);
+    }
+
+    if (di->arguments->size() < 1) {
+        ::error(ErrorType::ERR_MODEL, "Expected at least 1 argument: %1%", di);
+        return;
+    }
+
+    auto declaredSize = di->arguments->at(0)->expression->to<IR::Constant>();
+    if (!declaredSize->fitsUint()) {
+        ::error(ErrorType::ERR_OVERLIMIT, "%1%: size too large", declaredSize);
+        return;
+    }
+    size = declaredSize->asUnsigned();
+
+    if (di->arguments->size() == 2) {
+        this->initialValue = di->arguments->at(1)->expression->to<IR::Constant>();
```

Comment:
Maybe you should document the fact that this can be nullptr.
But if the argument is present and not a constant you will also get a nullptr, which is probably wrong.